File: momentra/momentra/Shell/ShellPerf.swift
import Foundation
import os

/// S9-J lightweight shell performance marks (source parity with Android ShellPerf).
enum ShellPerf {
    private static let log = Logger(subsystem: "com.momentra", category: "perf")
    private static var lastStore: [String: Int64] = [:]

    struct Mark {
        let name: String
        let startedAt: Date
        init(name: String, startedAt: Date = Date()) {
            self.name = name
            self.startedAt = startedAt
        }
    }

    static func start(_ name: String) -> Mark { Mark(name: name) }

    @discardableResult
    static func end(_ mark: Mark, extras: [String: Any] = [:]) -> Int64 {
        let elapsed = Int64(Date().timeIntervalSince(mark.startedAt) * 1000)
        var parts = ["event=\(mark.name)", "elapsedMs=\(elapsed)"]
        for (k, v) in extras {
            parts.append("\(k)=\(v)")
        }
        log.info("\(parts.joined(separator: " "), privacy: .public)")
        lastStore[mark.name] = elapsed
        return elapsed
    }

    static func instant(_ name: String, extras: [String: Any] = [:]) {
        _ = end(Mark(name: name, startedAt: Date()), extras: extras)
    }

    static var last: [String: Int64] { lastStore }

    static func clear() { lastStore.removeAll() }
}

File: apk/app/src/main/java/com/example/momentra/ui/shell/perf/ShellPerf.kt
package com.example.momentra.ui.shell.perf

/**
 * S9-J lightweight shell performance marks.
 * On device: android.util.Log. In JVM unit tests Log stubs throw — swallowed.
 */
object ShellPerf {
    private const val TAG = "MomentraPerf"

    data class Mark(
        val name: String,
        val startedAtMs: Long = System.currentTimeMillis(),
    )

    fun start(name: String): Mark = Mark(name)

    fun end(mark: Mark, extras: Map<String, Any?> = emptyMap()): Long {
        val elapsed = System.currentTimeMillis() - mark.startedAtMs
        val payload = buildString {
            append("event=").append(mark.name)
            append(" elapsedMs=").append(elapsed)
            extras.forEach { (k, v) ->
                if (v != null) append(' ').append(k).append('=').append(v)
            }
        }
        emit(payload)
        last[mark.name] = elapsed
        return elapsed
    }

    fun instant(name: String, extras: Map<String, Any?> = emptyMap()) {
        end(Mark(name, System.currentTimeMillis()), extras)
    }

    val last: MutableMap<String, Long> = mutableMapOf()

    fun clear() {
        last.clear()
    }

    private fun emit(payload: String) {
        try {
            android.util.Log.i(TAG, payload)
        } catch (_: Throwable) {
            // Unit-test JVM stubs throw for Log.*; marks still recorded in `last`.
        }
    }
}
